service/stream/exec_data_streamer.go

```diff
@@ -0,0 +1,168 @@
+		accessApi: accessAPI,
+		queue:     archive.NewDeque(),
+		buffer:    archive.NewDeque(),
+		limit:     cfg.BufferSize,
+		busy:      0,
+		ctx:       ctx,
+		chain:     chain,
+	}
+}
+
+// OnBlockFinalized is a callback for the Flow consensus follower. It is called
+// each time a block is finalized by the Flow consensus algorithm.
+func (e *ExecDataStreamer) OnBlockFinalized(block *model.Block) {
+	blockID := block.BlockID
+	// We push the block ID to the front of the queue; the streamer will try to
+	// download the blocks in a FIFO manner.
+	e.queue.PushFront(blockID)
+
+	e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for download")
+}
+
+// Next returns the next available block data. It returns an ErrUnavailable if no block
+// data is available at the moment.
+func (e *ExecDataStreamer) Next() (*uploader.BlockData, error) {
+	// same implementation as GCPStreamer
+	go e.poll()
+	if e.buffer.Len() == 0 {
+		e.log.Debug().Msg("buffer empty, no execution record available")
+		return nil, archive.ErrUnavailable
+	}
+	record := e.buffer.PopBack()
+	return record.(*uploader.BlockData), nil
+}
+
+func (e *ExecDataStreamer) poll() {
+	// same implementation as GCPStreamer
+	if !atomic.CompareAndSwapUint32(&e.busy, 0, 1) {
+		return
+	}
+	defer atomic.StoreUint32(&e.busy, 0)
+	err := e.pullExecData()
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		e.log.Debug().Msg("next execution record not available, download stopped")
+		return
+	}
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not download execution records")
+		return
+	}
+}
+
+func (e ExecDataStreamer) pullExecData() error {
+	for {
+		// same implementation as GCPStreamer
+		if uint(e.buffer.Len()) >= e.limit {
+			e.log.Debug().Uint("limit", e.limit).Msg("buffer full, stopping execution record download")
+			return nil
+		}
+
+		if uint(e.queue.Len()) == 0 {
+			e.log.Debug().Msg("queue empty, stopping execution record download")
+			return nil
+		}
+		blockID := e.queue.PopBack().(flow.Identifier)
+		record, err := e.getUploaderBlockData(blockID)
+		if err != nil {
+			e.queue.PushBack(blockID)
+			return fmt.Errorf("could not pull execution record (name: %s): %w", blockID, err)
+		}
+
+		e.log.Debug().
+			Str("name", blockID.String()).
+			Uint64("height", record.Block.Header.Height).
+			Hex("block", blockID[:]).
+			Msg("pushing execution record into buffer")
+
+		e.buffer.PushFront(record)
+	}
+}
+
+func (e *ExecDataStreamer) getUploaderBlockData(blockID flow.Identifier) (*uploader.BlockData, error) {
+	// TODO : get rid of uploader.BlockData use
+	// we currently have to query additional data from ANs to keep a 1:1 match of the data from the GCP streamer
+	var errs *multierror.Error
+	// get block
+	br := &access.GetBlockByIDRequest{Id: blockID[:]}
+	b, err := e.accessApi.GetBlockByID(e.ctx, br)
+	if err != nil {
+		errs = multierror.Append(errs, fmt.Errorf("failed to get block data for blockID (%s): %w", blockID, err))
+	}
```

Comment:
could we just get this from the local storage?

Reply:
we'd only have the header in local

Reply:
why only the header? the archive node is running a consensus follower which should be indexing full blocks